src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11KDF.java

```diff
@@ -0,0 +1,356 @@
+            throws InvalidAlgorithmParameterException,
+            NoSuchAlgorithmException {
+        if (alg == null) {
+            throw new NullPointerException("the algorithm for the " +
+                    "SecretKey return value must not be null");
+        }
+        if (alg.isEmpty()) {
+            throw new NoSuchAlgorithmException("the algorithm for the " +
+                    "SecretKey return value must not be empty");
+        }
+        return derive(alg, derivationSpec, SecretKey.class);
+    }
+
+    @Override
+    protected byte[] engineDeriveData(AlgorithmParameterSpec derivationSpec)
+            throws InvalidAlgorithmParameterException {
+        return derive("Generic", derivationSpec, byte[].class);
+    }
+
+    private <T> T derive(String alg, AlgorithmParameterSpec derivationSpec,
+            Class<T> retType) throws InvalidAlgorithmParameterException {
+        SecretKey baseKey;
+        SecretKey salt = null;
+        byte[] info = null;
+        int outLen;
+        boolean isExtract = false, isExpand = false;
+        boolean isData = retType == byte[].class;
+        assert isData || retType == SecretKey.class : "Invalid return type.";
+        assert alg != null : "The algorithm cannot be null.";
+
+        switch (derivationSpec) {
+            case HKDFParameterSpec.Extract anExtract -> {
+                isExtract = true;
+                baseKey = consolidateKeyMaterial(anExtract.ikms());
+                salt = consolidateKeyMaterial(anExtract.salts());
+                outLen = svcKi.prkLen / 8;
+                assert outLen * 8 == svcKi.prkLen : "Invalid PRK length.";
+            }
+            case HKDFParameterSpec.Expand anExpand -> {
+                isExpand = true;
+                baseKey = anExpand.prk();
+                outLen = anExpand.length();
+                info = anExpand.info();
+            }
+            case HKDFParameterSpec.ExtractThenExpand anExtractExpand -> {
+                isExtract = true;
+                isExpand = true;
+                baseKey = consolidateKeyMaterial(anExtractExpand.ikms());
+                salt = consolidateKeyMaterial(anExtractExpand.salts());
+                outLen = anExtractExpand.length();
+                info = anExtractExpand.info();
+            }
+            case null -> throw new NullPointerException(
+                    "derivationSpec must be a " + HKDFParameterSpec.class +
+                    " instance, instead of null.");
+            default -> throw new InvalidAlgorithmParameterException(
+                    "derivationSpec must be a " + HKDFParameterSpec.class +
+                    " instance, instead of " + derivationSpec.getClass());
+        }
+
+        P11Key p11BaseKey = convertKey(baseKey, (isExtract ? "IKM" : "PRK") +
+                " could not be converted to a token key for HKDF derivation.");
+
+        long saltType = CKF_HKDF_SALT_NULL;
+        byte[] saltBytes = null;
+        P11Key p11SaltKey = null;
```

Comment:
For `Expand`, `salt` is null.

Reply:
For Expand, `salt` should be `null`. Do you mean adding or hitting an assertion?

Reply:
Assertion on line 162 will throw an NPE.

Reply:
Ah, you're right, the assertion throws an NPE. Fixed.